### PR TITLE
* Products witohut initial cost

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/11094

### CODE REVIEW

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            )
+AS
+$$
+WITH element AS
+         (
+             SELECT el.m_costelement_id
+
+             FROM c_acctschema s
+                      JOIN m_costelement el on s.costingmethod = el.costingmethod
+             WHERE s.ad_org_id = 1000000
```
Comment (metas-ts) [inline backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql:19]:
```suggestion
             WHERE s.ad_org_id = p_ad_org_id
```

..but please also check my private remark

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql (hunk 1 above)
Comment (metas-rc) [reply backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql]: I forgot my testing data in there.  Fixed now.

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql [hunk 2]
```diff
@@ -0,0 +1,80 @@
+DROP FUNCTION IF EXISTS ProductsWithoutInitialCost
+(p_D_Client_ID NUMERIC,
+ p_ad_org_id NUMERIC);
+
+CREATE OR REPLACE FUNCTION ProductsWithoutInitialCost(
```
Comment (teosarca) [inline backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql:5]: why not `de_metas_acct.report_ProductsWithoutInitialCost`?

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                      LEFT JOIN firstCostDetails fcd on p.m_product_id = fcd.m_product_id
+                      LEFT JOIN m_costdetail cd on fcd.firstCostDetail = cd.m_costdetail_id
+             WHERE cd.m_costdetail_id IS NULL
+                OR cd.issotrx = 'N'
+         )
+SELECT fp.AD_Org_ID,
+       o.value as OrgValue,
+       o.name  as OrgName,
+       p.M_Product_ID,
+       p.value,
```
Comment (teosarca) [inline backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql:59]: as ProductValue

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+       p.name
```
Comment (teosarca) [inline backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql:60]: as ProductName

File: backend/de.metas.acct.base/src/main/sql/postgresql/system/43-de.metas.acct/5588320_sys_gh11094_Products_Winthout_Initial_Cost_Function.sql [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                      JOIN m_costelement el on s.costingmethod = el.costingmethod
+         ),
+     firstCostDetails AS
+         (
+             SELECT cd.M_Product_ID,
+                    cd.ad_org_id,
+                    min(M_CostDetail_ID) as firstCostDetail
+
+             from M_costDetail cd
+                      JOIN element on cd.m_costelement_id = element.m_costelement_id
```
Comment (teosarca) [inline backend/de.metas.acct.base/src/main/sql/postgresql/system/43-de.metas.acct/5588320_sys_gh11094_Products_Winthout_Initial_Cost_Function.sql:35]: here u shall filter by M_CostDetail.C_AcctSchema_ID too

File: backend/de.metas.acct.base/src/main/sql/postgresql/system/43-de.metas.acct/5588320_sys_gh11094_Products_Winthout_Initial_Cost_Function.sql [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+AS
+$$
+WITH element AS
+         (
+             SELECT el.m_costelement_id
+
+             FROM c_acctschema s
+                      JOIN ad_clientinfo ci on ci.ad_client_id = p_AD_Client_Id
+                 AND s.c_acctschema_id = ci.c_acctschema1_id
+                      JOIN m_costelement el on s.costingmethod = el.costingmethod
```
Comment (teosarca) [inline backend/de.metas.acct.base/src/main/sql/postgresql/system/43-de.metas.acct/5588320_sys_gh11094_Products_Winthout_Initial_Cost_Function.sql:26]: I would introduce the p_M_CostElement_ID parameter.
Is there any reason to not do it?

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.acct.base/src/main/sql/postgresql/system/43-de.metas.acct/5588320_sys_gh11094_Products_Winthout_Initial_Cost_Function.sql (hunk 6 above)
Comment (metas-rc) [reply backend/de.metas.acct.base/src/main/sql/postgresql/system/43-de.metas.acct/5588320_sys_gh11094_Products_Winthout_Initial_Cost_Function.sql]: > I would introduce the p_M_CostElement_ID parameter.
> Is there any reason to not do it?

Added it and made it mandatory. The default value will be the one with the same costing method as the one form the schema in the client info.

File: backend/de.metas.acct.base/src/main/sql/postgresql/system/43-de.metas.acct/5588320_sys_gh11094_Products_Winthout_Initial_Cost_Function.sql (hunk 5 above)
Comment (metas-rc) [reply backend/de.metas.acct.base/src/main/sql/postgresql/system/43-de.metas.acct/5588320_sys_gh11094_Products_Winthout_Initial_Cost_Function.sql]: done

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql (hunk 3 above)
Comment (metas-rc) [reply backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql]: done

File: backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql (hunk 4 above)
Comment (metas-rc) [reply backend/de.metas.acct.base/src/main/sql/postgresql/ddl/functions/ProductsWithoutInitialCost.sql]: done

**Pull request merged**